core/order.go

```diff
@@ -252,6 +292,237 @@ func (n *OpenBazaarNode) Purchase(data *PurchaseData) error {
+		}
+		satoshis, err := n.getPriceInSatoshi(l.Item.Price)
+		if err != nil {
+			return 0, err
+		}
+		itemTotal += satoshis
+		for _, option := range item.Options {
+			optionExists := false
+			for _, listingOption := range l.Item.Options {
+				if strings.ToLower(option.Name) == strings.ToLower(listingOption.Name) {
+					optionExists = true
+					variantExists := false
+					for _, variant := range listingOption.Variants {
+						if strings.ToLower(variant.Name) == strings.ToLower(option.Value) {
+							if variant.PriceModifier != nil {
+								satoshis, err := n.getPriceInSatoshi(variant.PriceModifier)
+								if err != nil {
+									return 0, err
+								}
+								itemTotal += satoshis
+							}
+							variantExists = true
+							break
+						}
+					}
+					if !variantExists {
+						return 0, errors.New("Selected variant not found in listing")
+					}
+					break
+				}
+			}
+			if !optionExists {
+				return 0, errors.New("Selected option not found in listing")
+			}
+		}
+		//TODO: should apply coupon discounts to itemTotal here
+		itemTotal *= uint64(item.Quantity)
+		total += itemTotal
+	}
+
+	// Add in shipping costs
+	type combinedShipping struct {
+		quantity int
+		price    uint64
+		add      bool
+		modifier uint64
+	}
+	var combinedOptions []combinedShipping
+
+	var shippingTotal uint64
+	for listingHash, listing := range physicalGoods {
+		for _, item := range contract.BuyerOrder.Items {
+			if item.ListingHash == listingHash {
+				var itemShipping uint64
+				// Check selected option exists
+				var option *pb.Listing_ShippingOption
+				for _, shippingOption := range listing.ShippingOptions {
+					if shippingOption.Name == item.ShippingOption.Name {
+						option = shippingOption
+						break
+					}
+				}
+				if option == nil {
+					return 0, errors.New("Shipping option not found in listing")
+				}
+
+				// Check that this option ships to us
+				shipsToMe := false
+				for _, country := range option.Regions {
+					if country == contract.BuyerOrder.Shipping.Country || country == pb.CountryCode_ALL {
+						shipsToMe = true
+						break
+					}
+				}
+				if !shipsToMe {
+					return 0, errors.New("Listing does ship to selected country")
+				}
+
+				// Check service exists
+				var service *pb.Listing_ShippingOption_Service
+				for _, shippingService := range option.Services {
+					if strings.ToLower(shippingService.Name) == strings.ToLower(item.ShippingOption.Service) {
+						service = shippingService
+					}
+				}
+				if service == nil {
+					return 0, errors.New("Shipping service not found in listing")
+				}
+				shippingSatoshi, err := n.getPriceInSatoshi(service.Price)
+				if err != nil {
+					return 0, err
+				}
+				shippingPrice := uint64(item.Quantity) * shippingSatoshi
+				itemShipping += shippingPrice
+
+				// Apply shipping rules
+				if option.ShippingRules != nil {
+					for _, rule := range option.ShippingRules.Rules {
+						if int(option.ShippingRules.RuleType) == 0 {
```

Comment:
@cpacia A switch is a bit more idiomatic but that's purely stylistic. Not a major issue.